src/Components/Server/src/Circuits/CircuitPrerenderer.cs

```diff
@@ -2,47 +2,77 @@
+        public async Task<ComponentPrerenderResult> PrerenderComponentAsync(ComponentPrerenderingContext prerenderingContext)
         {
             var context = prerenderingContext.Context;
-            var circuitHost = _circuitFactory.CreateCircuitHost(
-                context,
-                client: CircuitClientProxy.OfflineClient,
-                GetFullUri(context.Request),
-                GetFullBaseUri(context.Request));
-
-            // We don't need to unsubscribe because the circuit host object is scoped to this call.
-            circuitHost.UnhandledException += CircuitHost_UnhandledException;
-
-            // For right now we just do prerendering and dispose the circuit. In the future we will keep the circuit around and
-            // reconnect to it from the ComponentsHub. If we keep the circuit/renderer we also need to unsubscribe this error
-            // handler.
-            try
+            var circuitHost = GetOrCreateCircuitHost(context);
+
+            var renderResult = await circuitHost.PrerenderComponentAsync(
+                prerenderingContext.ComponentType,
+                prerenderingContext.Parameters);
+
+            circuitHost.Descriptors.Add(new ComponentDescriptor
             {
-                return await circuitHost.PrerenderComponentAsync(
-                    prerenderingContext.ComponentType,
-                    prerenderingContext.Parameters);
+                ComponentType = prerenderingContext.ComponentType,
+                Prerendered = true
+            });
+
+            var result = new[] {
+                $"<!-- M.A.C.Component:{{\"circuitId\":\"{circuitHost.CircuitId}\",\"rendererId\":\"{circuitHost.Renderer.Id}\",\"componentId\":\"{renderResult.ComponentId}\"}} -->",
+            }.Concat(renderResult.Tokens).Concat(
+                new[] {
+                    $"<!-- M.A.C.Component: {renderResult.ComponentId} -->"
+                });
+
+            return new ComponentPrerenderResult(result);
+        }
+
+        private CircuitHost GetOrCreateCircuitHost(HttpContext context)
+        {
+            if (context.Items.TryGetValue(CircuitHostKey, out var existingHost))
+            {
+                return (CircuitHost)existingHost;
             }
-            finally
+            else
             {
-                await circuitHost.DisposeAsync();
+                var result = _circuitFactory.CreateCircuitHost(
+                    context,
+                    client: CircuitClientProxy.CreateOffline(),
+                    GetFullUri(context.Request),
+                    GetFullBaseUri(context.Request));
+
+                result.UnhandledException += CircuitHost_UnhandledException;
+                context.Response.OnCompleted(() =>
+                {
+                    result.UnhandledException -= CircuitHost_UnhandledException;
+                    return Task.CompletedTask;
+                });
+                _registry.RegisterDisconectedCircuit(result);
```

Comment:
There's a subtle race here that might not be possible to hit in practice. The circuit is going to be registered as a *disconnected* circuit while it still might be used by the request.

Reply:
Yeah, but I don't expect us to hit it. Alternatively we could register a callback for OnRequestCompleted and perform the registration in there.

Reply:
I changed my mind. It's free to do it in OnResponseCompleted and more correct.